File: 프로그래머스/2/135807. 숫자 카드 나누기/숫자 카드 나누기.java
import java.util.*;
class Solution {
    public int solution(int[] arrayA, int[] arrayB) {
        int gcdA = arrayA[0];
        int gcdB = arrayB[0];
        for(int i = 1; i<arrayA.length; i++){
            gcdA = gcd(gcdA, arrayA[i]);
            gcdB = gcd(gcdB, arrayB[i]);
        }
        
        int answerA = 0;
        int answerB = 0;
        if(canDivide(gcdA, arrayB)) answerA = gcdA;
        if(canDivide(gcdB, arrayA)) answerB = gcdB;
        
        
        
        return Math.max(answerA, answerB);
    }
    
    static int gcd(int a, int b){
        while(b != 0){
            int temp = a % b;
            a = b;
            b = temp;
        }
        return a;
    }
    
    static boolean canDivide(int num, int[] arr){
        for(int a: arr){
            if(a % num == 0) return false;
        }
        return true;
    }
}

File: 프로그래머스/2/135807. 숫자 카드 나누기/숫자 카드 나누기.kt
class Solution {
    fun solution(arrayA: IntArray, arrayB: IntArray): Int {
        val gcdA = getgcdArray(arrayA)
        val gcdB = getgcdArray(arrayB)
        
        val answerA = if (canDivide(gcdA, arrayB)) gcdA else 0
        val answerB = if (canDivide(gcdB, arrayA)) gcdB else 0
        
        return maxOf(answerA, answerB)
    }
    
    private fun getgcdArray(arr: IntArray): Int {
        var gcd = arr[0]
        for(i in 1 until arr.size) {
            gcd = gcd(gcd, arr[i])
        }
        return gcd
    }
    
    private fun gcd(a: Int, b: Int): Int{
        var x = a
        var y = b
        while(y != 0){
            val temp = x % y
            x = y
            y = temp
        }
        return x
    }
    
    private fun canDivide(num: Int, arr: IntArray): Boolean {
        for(x in arr){
            if(x % num == 0) return false
        }
        return true
    }
}